### PR TITLE
YOLO Dataset Experiments: Ablation, Balancing, and Preprocessing (Roboflow)

### LINKED ISSUE #15: YOLO Dataset Experiments: Ablation, Balancing, and Preprocessing (Roboflow)
# Task Overview:
Your task is to create multiple versions of the combined dataset by applying at least one dataset improvement technique, training one lightweight YOLO model per dataset variant using Roboflow. The purpose is to understand how dataset quality and composition affect model performance.

**Each pair should:**
1. Make a copy of the [combined dataset](https://app.roboflow.com/caretech-qnw0y/v1-caretech-combined-dataset/1/export) (if you haven't already done so)
2. Apply one or more of the techniques listed below
3. Train one lightweight YOLO model in Roboflow for each modified dataset that you create
4. Record results and observations

**Dataset Techniques:**
- Dataset ablation
  - This is an umbrella term but in this case we are suggesting to **remove** ~10% of images (by class). Feel free to choose a different percentage if you wish.
- Class balancing
  - Reduce overrepresented classes and/or augment underrepresented ones
    - **Check if there are any YOLO hyperparameters that can help you do this!
- Image quality preprocessing
  - Detect dark or low-contrast images (e.g., Laplacian variance)
  - Brighten or enhance those images before training
  - **Check if there are any YOLO hyperparameters that can help you do this!

**Working with Roboflow to Upload New Datasets** 
- For each dataset variation that you create, please upload it to a new dataset / project on Roboflow
- Make sure you create a new workspace with your partner on Roboflow with the naming format: `caretech-partner1-partner2`
- Add each dataset / project to your respective workspace on Roboflow
- You can find more info and snippets related to Roboflow dataset uploading and more on the [CareTech Resources](https://docs.google.com/document/d/1xWkiNY6oqvcNrNZ_sTFa7KFGUolHKp4kYOogBnB2Wso/edit?tab=t.sf58t38j3czj)
- NOTE: uploading a dataset to Roboflow may take some time

**Model Training Requirements**
- Train using Roboflow! Since Google CoLab has credit limits, once you are done with the dataset you would like to train, upload it to a **Roboflow Workspace** and train it in Roboflow. What is nice about Roboflow is that once you upload the dataset and choose which model to run, you can leave it there and it will email you when it is finished.
- Choose a lightweight YOLO model (e.g., YOLOv8-nano or YOLOv8-small)

**What to Record:**
- [ ] Which technique(s) you applied
- [ ] Dataset size and class distribution (before vs after)
- [ ] YOLO model used
- [ ] Validation metrics (mAP, precision, recall)
- [ ] 1–2 sentences on what you observed
- [ ] Link to your Roboflow workspace with your uploaded datasets

## Important Reminders:
- Follow proper naming conventions of **files** and **directories**
- Include your work in separate cells and **delete** your work by changing it and running it again (unless you would like that specific code's output not documented)
- Include your combined dataset in the .gitignore as the proper 'v1-combined-dataset' as if each pair changes the name, it makes the .gitignore unnecessarily long
- Any files or folders you make should be in your respective pair folder
- Provide context to what you're trying to do with the code
- If there are visualizations created, provide insight as to what it tells you

If you have any questions, please reach out to Leo and Margaret! We hope that the time can be more accomodating and Roboflow will be easier to work with than Colab. Thank you so much for letting us know about your issues and best of luck with training your new model!

### PR DESCRIPTION
**Pull request by audreyyluu**
Model used: YOLO26 Fast, 70 epochs

[caretech-audrey-lahari roboflow workspace](https://app.roboflow.com/caretech-audrey-lahari?skipSetup=true)

Techniques tested
- Used 10% dataset ablation targeting the majority classes
  - mAP50: 68.6%
  - Precision: 65.3%
  - Recall: 63.4%
- Attempted class balancing by upsampling/augmenting the minority classes
   - Brightness, saturation, hue, blur, rotation, horizontal flip, shear (more on these down below)
   - Increased the minority class images by 3x
- Augmentation increased precision and decreased recall
- Contrast enhancement (decreased mAP, precision, and recall)

Observations
- Miso soup, orange, and meatloaf are being frequently missed
- Lau lau is often detected when it is not there (518 false detections)

Lahari:
For my augmentation choices, I referenced the Ultralytics documentation for YOLO data augmentation, which has tips on when to use each augmentation choice. 
- I used brightness to train the model to get used to the same object showing up in different lighting. A percentage of 10-40% was recommended, so I chose 20%. 
- I used saturation to train the model to get used to the same object being taken from different camera settings, which might each have different amounts of saturation. Chose 10% assuming real-life food images don't change in saturation more than ~10%. Looking back in the future I can try increasing the saturation.
- I used hue for the same reasons as brightness and saturation. Chose 5% so that the model is not overloaded with too many similar augmentations.
- I used blur to train the model to get used to the same object being shown in different image qualities/bluriness. Chose 10% assuming real-life food images don't change in blur more than ~10%. Looking back in the future I can try increasing the blur.
- I used rotation, horizontal flip, and shear to train the model to get used to the same object being shown in different angles. Generally a rotation of 10% is common, so I chose 7% to not start off too strong. Shear values can rapidly distort the image, so the documentation recommended to start with small values, so I chose 5%.

Part of #15

### CODE REVIEW

**Review by margoglvz — CHANGES_REQUESTED** [review]
Hi guys thank you for your PR! I really enjoyed seeing how you guys focused on  improving the minority classes and testing different preprocessing techniques. Here is some feedback and I did have some **questions** you guys can answer when you have the time.

- Remove any unecessary files before making your PR
  - I believe there was also a temp file in another PR as well so please make sure to **remove** any files you don't want to be reviewed or need in your PR
- Include more accurate and descriptive file names
  - Instead of 'lahari_main.ipynb' --> 'ablation_augmentation_preprocessing.ipynb'
 - **Question:** In your code block of augmentation on minority classes, (feel free to correct me if I am wrong) is it removing the bounding box label of the majority classes in an image and then making a copy? So, if there is a minority and a majority class food in the same image, it only removes the majority class bounding box info from `labels`, and then copying it? I see that the 'apple' class went from 36 images to 91 images after running this code so I am assuming that it makes exact copies.
   - I am not sure how much this will improve the model as even though the majority class label is not there anymore, the model is still **seeing** the majority food class object in the background. I also see that you did do some classic data augmentation in roboflow (rotation, hue, etc.) but I think it might be better to simply do the augmentations than removing the bounding box labels of majority classes. Also, if the images were copied (for minority classes), that would **not** affect model performance since it is the same image. In order for it to possibly improve the model, it has to be **modified** in some way (exact copies won't affect performance). In summary, if you would like to add augmentation in Roboflow, there is no need to make exact copies of the minority classes.
- **Question**: Could you also provide a little reasoning as to why you decided to go with your augmentation choices? As that was part of the PR requirements. Also, how did you guys decide the strength of your filters? For example, for the contrast filter and hue filter, how did you decide how strong you wanted it to be? All the filters might not work well for **all** images so it is important that that is checked.

The next task we give you all on Wednesday will build on top of this for more model improvements! So definitely wanna make sure we learn what we should try out next to improve the models. Thank you so much and let me know if you have any questions!
